File: official-docs/examples/examples/python-server-wallets/server_wallet/evm/sign_message.py
"""
Dynamic Message Signing Demo

Sign messages with Dynamic server wallets for authentication,
verification, and proof of ownership.

## Usage

  python -m server_wallet.evm.sign_message "Hello, World!"
  python -m server_wallet.evm.sign_message "Hello, World!" --address 0x123...
  python -m server_wallet.evm.sign_message "Hello, World!" --address 0x123... --password xyz

## Use Cases

- Authenticate users by proving wallet ownership
- Sign authorization tokens or session data
- Verify identity without on-chain transactions
- Create off-chain signatures for gasless flows
"""

import argparse
from datetime import datetime

from lib.cli import run_script
from lib.dynamic import authenticated_evm_client
from lib.wallet_helpers import get_or_create_evm_wallet


async def sign_message(message: str, address: str | None = None, password: str | None = None):
    async with await authenticated_evm_client() as client:
        # Step 1: Get or create wallet
        wallet_address, key_shares = await get_or_create_evm_wallet(client, address, password)

        # Step 2: Sign the message using MPC threshold signatures
        print("\nSigning message...")
        start = datetime.now()

        signature = await client.sign_message(
            message=message,
            address=wallet_address,
            key_shares=key_shares,
        )

        duration = (datetime.now() - start).total_seconds()

        # Step 3: Display results
        print(f"\nMessage signed in {duration:.2f}s")
        print(f"Message: \"{message}\"")
        print(f"Signature: {signature}")
        print(f"Signer: {wallet_address}")

        return signature


def main():
    parser = argparse.ArgumentParser(description="Sign messages with Dynamic server wallets")
    parser.add_argument("message", help="Message to sign")
    parser.add_argument("--address", help="Use a saved wallet by address")
    parser.add_argument("--password", help="Password for password-protected wallets")

    args = parser.parse_args()

    run_script(lambda: sign_message(args.message, args.address, args.password))


if __name__ == "__main__":
    main()

File: official-docs/examples/examples/nodejs-server-wallets/src/server-wallet/sign-message.ts
#!/usr/bin/env tsx

/**
 * Dynamic Message Signing Demo
 *
 * Sign messages with Dynamic server wallets for authentication,
 * verification, and proof of ownership.
 *
 * ## Usage
 *
 *   pnpm sign-msg "Hello, World!"                                    # Sign with new ephemeral wallet
 *   pnpm sign-msg "Hello, World!" --address 0x123...                 # Sign with saved wallet
 *   pnpm sign-msg "Hello, World!" --address 0x123... --password xyz  # Sign with password-protected wallet
 *
 * ## Use Cases
 *
 * - Authenticate users by proving wallet ownership
 * - Sign authorization tokens or session data
 * - Verify identity without on-chain transactions
 * - Create off-chain signatures for gasless flows
 */

import { parseArgs, runScript } from "../lib/cli";
import { authenticatedEvmClient } from "../lib/dynamic";
import { getOrCreateWallet, type WalletInfo } from "../lib/wallet-helpers";

/**
 * Step 2: Sign a message with the wallet
 */
async function signMessage(
  message: string,
  wallet: WalletInfo,
  password?: string,
) {
  // Get authenticated Dynamic client
  const dynamicEvmClient = await authenticatedEvmClient();

  console.info(`\nSigning message...`);
  const start = Date.now();

  // Sign the message using the wallet's key shares
  // If key shares are available locally, pass them directly.
  // If empty (backed up to Dynamic), omit them so the SDK recovers from backup using the password.
  const signature = await dynamicEvmClient.signMessage({
    accountAddress: wallet.address,
    ...(wallet.externalServerKeyShares.length > 0 && {
      externalServerKeyShares: wallet.externalServerKeyShares,
    }),
    message,
    ...(password && { password }),
  });

  const duration = ((Date.now() - start) / 1000).toFixed(2);

  // Step 3: Display results
  console.info(`\nMessage signed in ${duration}s`);
  console.info(`Message: "${message}"`);
  console.info(`Signature: ${signature}`);
  console.info(`Signer: ${wallet.address}`);

  return signature;
}

function showUsage(): never {
  console.error("Please provide a message to sign");
  console.error("\nUsage:");
  console.error('  pnpm sign-msg "Hello, World!"');
  console.error('  pnpm sign-msg "Hello, World!" --address 0x123...');
  console.error(
    '  pnpm sign-msg "Hello, World!" --address 0x123... --password xyz',
  );
  process.exit(1);
}

runScript(async () => {
  const { positional, getFlag } = parseArgs(process.argv);

  // Parse arguments
  const message = positional[0];
  const address = getFlag("address");
  const password = getFlag("password");

  if (!message) {
    showUsage();
  }

  // Step 1: Get or create wallet
  const dynamicEvmClient = await authenticatedEvmClient();
  const wallet = await getOrCreateWallet(dynamicEvmClient, address, password);

  // Step 2: Sign the message
  await signMessage(message, wallet, password);
});
